### DISCUSSION TITLE
Handling Fallback for Custom LogEventFactory

### DISCUSSION THREAD

**Discussion by skjelmo** [apache/logging-log4j2#1798] [Q&A] [answered]
## Questions
- Will Log4J default to the default LogEventFactory if a custom LogEventFactory defined through the system property `Log4jLogEventFactory` is not available?
([Documentation for LogEventFactory](https://logging.apache.org/log4j/2.x/manual/extending.html#logeventfactory))
- Is it possible to define Log4J's built-in default LogEventFactory as a fallback? 

## Context
- We have configured our own LogEventFactory as [described in the documentation](https://logging.apache.org/log4j/2.x/manual/extending.html#logeventfactory). The intention is to split log statements for throwables with long stacktraces into separate log events to circumvent a [limitation in Heroku where log lines over 10 000 bytes are split into separate chunks](https://devcenter.heroku.com/articles/limits#line-length) resulting in invalid JSON format limiting log analysis and observability.
- We use the New Relic Java Agent which bundles Log4J as a dependency of it's own. It is also affected by the system property `Log4jLogEventFactory` when we configured Log4J to use our custom LogEventFactory, causing a `ClassNotFoundException` as our custom LogEventFactory is not available to the New Relic Java Agent. 
- I [raised this issue](https://forum.newrelic.com/s/hubtopic/aAX8W000000PHCwWAO/custom-log4j-logeventfactory-affecting-new-relic) first with New Relic support who believe it is up to Log4J how it handles if the LogEventFactory defined in the system property `Log4jLogEventFactory` is not available.

**Comment by ppkarwasz** [comment, marked as answer]
@skjelmo,

Yes, even if the error message fails to mention it, Log4j Core falls back to a default log event factory, if your custom factory is not available (cf. [source code](https://github.com/apache/logging-log4j2/blob/f2dd4465e9cc73dde0042ec0f275c6f68e52192f/log4j-core/src/main/java/org/apache/logging/log4j/core/config/LoggerConfig.java#L84C6-L101)).

If you want to prevent the message from appearing, you can bundle your custom `LogEventFactory` together with a `log4j2.component.properties` containing:
```
log4j2.logEventFactory = fully_qualified_name_of_your_class
```

**Remark**: a custom `LogEventFactory` is not necessarily the best tool for the job, since it can not split events. You can split events with a custom `Appender` implementation or you can limit the size of the stacktrace (each layout has a way to limit stacktraces).

**Comment by skjelmo** [comment]
Thank you very much for the answer @ppkarwasz!

A custom `Appender` implementation delegating to the existing `Appender` worked wonders, and I hadn't considered it as a possibility before your remark.
